feat(forms): implement setErrors

vsavkin opened the pull request:
Example:

```
var login = new Control("someLogin");
c.setErrors({"notUnique": true});
expect(c.valid).toEqual(false);
expect(c.errors).toEqual({"notUnique": true});

c.updateValue("newLogin");
expect(c.valid).toEqual(true);
```

BREAKING CHANGE:

Before:

ControlGroup.errors and ControlArray.errors returned the reduced values of their children controls' errors.

After:

ControlGroup.errors and ControlArray.errors return the errors of the group and array.
And ControlGroup.controlsErrors and ControlArray.controlsErrors return the reduce values of their children controls' errors.

Notes:
- I had to separate errors and controlsErrors. This is to provide the following guarantee: _A control group with invalid children is invalid._
- Calling setErrors on a control will update the validity of its ancestors.

mhevery commented on modules/angular2/src/core/forms/model.ts:
```diff
@@ -204,6 +206,16 @@ export class Control extends AbstractControl {
+   */
+  _updateValue() {}
+
+  setErrors(errors: {[key: string]: any}): void {
```
Can we add docs? It is easier to do now, then during a big docs push.

mhevery commented on modules/angular2/src/core/forms/model.ts:
```diff
@@ -258,6 +270,21 @@ export class ControlGroup extends AbstractControl {
     return c && this._included(controlName);
   }
 
+  setErrors(errors: {[key: string]: any}): void {
```
Docs?

mhevery commented on modules/angular2/src/core/forms/model.ts:
```diff
@@ -360,6 +387,21 @@ export class ControlArray extends AbstractControl {
    */
   get length(): number { return this.controls.length; }
 
+  setErrors(errors: {[key: string]: any}): void {
```
Docs?`

vsavkin committed:
feat(forms): Implement a way to manually set errors on a control

Example:

var login = new Control("someLogin");
c.setErrors({"notUnique": true});
expect(c.valid).toEqual(false);
expect(c.errors).toEqual({"notUnique": true});

c.updateValue("newLogin");
expect(c.valid).toEqual(true);

BREAKING CHANGE:

Before:

ControlGroup.errors and ControlArray.errors returned a reduced value of their children controls' errors.

After:

ControlGroup.errors and ControlArray.errors return the errors of the group and array.
And ControlGroup.controlsErrors and ControlArray.controlsErrors return the reduce value of their children controls' errors.

mary-poppins commented:
Merging PR #4917 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-4917](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-4917).

TedSander commented on modules/angular2/src/core/forms/model.ts:
```diff
@@ -364,6 +420,9 @@ export class ControlArray extends AbstractControl {
   _updateValue(): void { this._value = this.controls.map((control) => control.value); }
 
   /** @internal */
+  _calculateControlsErrors() { return Validators.array(this); }
```
I wonder if it makes sense to move group, and array from Validators? I'm worried that a user might try to pass them in the constructor, or pass them in as a combined validator. Should these functions be used as a regular validator at this point?

TedSander commented on modules/angular2/src/core/forms/model.ts:
```diff
@@ -151,7 +179,23 @@ export class AbstractControl {
+  }
+
+  private _calculateStatus(): string {
+    return isPresent(this._errors) || isPresent(this._controlsErrors) ? INVALID : VALID;
```
Just making a Note: When async validators are added this will need to get more complicated. Probably something like:
- any errors: return INVALD
- any pending controls: return PENDING
- else VALID

TedSander commented on modules/angular2/test/core/forms/model_spec.ts:
```diff
@@ -151,6 +151,61 @@ export function main() {
+          expect(g.errors).toEqual({"someGroupError": true});
+        });
+
+        it("update a value should reset errosr", () => {
```
s/errosr/errors

TedSander commented:
LGTM I like it turned out much cleaner than the original design.

Krustie101 commented:
I notice the following things in the source code:

- The getError method returns null if the path is empty, instead of looking in the errors map. Is this intentional?

-  The group/array validator only take into account the errors, not the controlErrors of the abstract controls in the group/array. Since the group/array validator are used to compute the controlErrors of the group/array, I expect the validy status of the group to be incorrect in the following case:  
control `A` in controlGroup `B` in controlGroup `C`  
Assume control `A` is invalid (either by internal validation or manual validation) and that we do not set the errors object of either `B` or `C`.
The controlGroup `B` is invalid because the controlErrors  object of `B` contains the error of `A`, but controlGroup `C` is valid because the error of `A` is not "passed through" to `C`.
